includes/multi-currency/class-currency-switcher-widget.php

```diff
@@ -0,0 +1,172 @@
+<?php
+/**
+ * WooCommerce Payments Currency Switcher Widget
+ *
+ * @package WooCommerce\Payments
+ */
+
+namespace WCPay\Multi_Currency;
+
+use WP_Widget;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Currency Switcher Widget Class
+ */
+class Currency_Switcher_Widget extends WP_Widget {
+
+	const DEFAULT_SETTINGS = [
+		'title'  => '',
+		'symbol' => true,
+		'flag'   => false,
+	];
+
+	/**
+	 * Register widget with WordPress.
+	 */
+	public function __construct() {
+		parent::__construct(
+			'currency_switcher_widget',
+			__( 'Currency Switcher', 'woocommerce-payments' ),
+			[ 'description' => __( 'Let your customers switch between your enabled currencies', 'woocommerce-payments' ) ]
+		);
+	}
+
+	/**
+	 * Front-end display of widget.
+	 *
+	 * @param array $args     Widget arguments.
+	 * @param array $instance Saved values from database.
+	 */
+	public function widget( $args, $instance ) {
+		$instance = wp_parse_args(
+			$instance,
+			self::DEFAULT_SETTINGS
+		);
+
+		$title = apply_filters( 'widget_title', $instance['title'] );
+
+		echo $args['before_widget']; // phpcs:ignore WordPress.Security.EscapeOutput
+		if ( ! empty( $title ) ) {
+			echo $args['before_title'] . $title . $args['after_title']; // phpcs:ignore WordPress.Security.EscapeOutput
+		}
+
+		?>
+		<form>
+			<select
+				name="currency"
+				aria-label="<?php echo esc_attr( $title ); ?>"
+				onchange="this.form.submit()"
+			>
+				<?php
+				foreach ( Multi_Currency::instance()->get_enabled_currencies() as $currency ) {
```

Comment:
What do you think of using dependency injection here to pass the `Multi_Currency` instance and make this more testable? That way we would be able to easily write some unit tests mocking `Multi_Currency`'s `get_enabled_currencies` and `get_selected_currencies` asserting the widget's behavior.

I think we'd be able to do that by adding a parameter to the `Currency_Switcher_Widget`'s constructor, instantiating it in the `Multi_Currency` class, and using that object in `register_widget`. Something like this (not tested or cleaned up 😄):

```
class Multi_Currency {
	...
	public function init() {
		...
		include_once WCPAY_ABSPATH . 'includes/multi-currency/class-currency-switcher-widget.php';

		add_action(
			'widgets_init',
			function () {
				register_widget( new Currency_Switcher_Widget( $this ) );
			}
		);
		...
```

Also, sorry for the drive-by review comment. I'll do a proper review on Monday.

Reply:
💯 Agree. I saw your code in #1860 and looks way better for a testable perspective.
Tests are my weak point, I've experience doing tests only in JS. Thanks for the tip! 🙂